core/src/main/java/hudson/PluginFirstClassLoader2.java

```diff
@@ -0,0 +1,167 @@
+package hudson;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import hudson.util.CompoundEnumeration;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Objects;
+import jenkins.util.URLClassLoader2;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Class loader that consults the system class loader, the plugin's {@code WEB-INF/lib/*.jar} and
+ * {@code WEB-INF/classes} directories, and the Jenkins core class loader (in that order).
+ *
+ * <p>To use this class loader, set {@code pluginFirstClassLoader} to {@code true} in the {@code
+ * maven-hpi-plugin} configuration.
+ *
+ * @author Basil Crow
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class PluginFirstClassLoader2 extends URLClassLoader2 {
+    static {
+        registerAsParallelCapable();
+    }
+
+    @CheckForNull private final ClassLoader systemClassLoader;
+
+    public PluginFirstClassLoader2(URL[] urls, ClassLoader parent) {
+        super(Objects.requireNonNull(urls), Objects.requireNonNull(parent));
+        systemClassLoader = getSystemClassLoader();
+    }
+
+    /**
+     * Load the class with the specified binary name. This method searches for classes in the
+     * following order:
+     *
+     * <ol>
+     *   <li>
+     *       <p>Invoke {@link #findLoadedClass(String)} to check if the class has already been
+     *       loaded.
+     *   <li>
+     *       <p>Invoke {@link #loadClass(String)} on the system class loader.
+     *   <li>
+     *       <p>Invoke {@link #findClass(String)} to find the class.
+     *   <li>
+     *       <p>Invoke {@link #loadClass(String)} on the parent class loader.
+     * </ol>
+     *
+     * <p>If the class was found using the above steps and the {@code resolve} flag is true, this
+     * method will then invoke {@link #resolveClass(Class)} on the resulting {@link Class} object.
+     *
+     * <p>This method synchronizes on the result of {@link #getClassLoadingLock(String)} during the
+     * entire class loading process.
+     *
+     * @param name The binary name of the class
+     * @param resolve If {@code true} then resolve the class
+     * @return The resulting {@link Class} object
+     * @throws ClassNotFoundException If the class could not be found
+     */
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        synchronized (getClassLoadingLock(name)) {
+            Class<?> c = findLoadedClass(name);
+            if (c == null && systemClassLoader != null) {
+                try {
+                    c = systemClassLoader.loadClass(name);
+                } catch (ClassNotFoundException e) {
+                    // ignore
+                }
+            }
```

Comment:
is this the correct order for plugin-first? (I beleive this is different to the old ANTClassLoader`` which not directly consult the system classloader, but leaves it to the parent (after it has tried to load from the plugin)?

https://github.com/apache/ant/blob/002596a9bb830257250ea394a2637d57209edb3b/src/main/org/apache/tools/ant/AntClassLoader.java#L1116-L1127

Reply:
I removed any references to the system class loader.